### PR TITLE
feat: support RDP connection to private IP machines

### LINKED ISSUE #32: Allow to connect to machines with private IP only
Some machines are only inside Intranet and don't have public IP. We want to connect to them from Internet. After supporting this, Casvisor will reach the same functionality as professional remote desktop software like TeamViewer, RustDesk, etc.

About how to implement it, I think an agent can be installed into the machine to get this issue done. It may work in a similar way like frp: https://github.com/fatedier/frp

### PR DESCRIPTION
**Pull request by leo220yuyaodog**
Fix: https://github.com/casvisor/casvisor/issues/32

### CODE REVIEW

File: object/asset.go [hunk 1]
```diff
@@ -218,7 +239,10 @@ func DeleteAsset(asset *Asset) (bool, error) {
 	return affected != 0, nil
 }
 
-func (asset *Asset) getId() string {
+func (asset *Asset) GetId() string {
+	if asset == nil {
```
Comment (hsluoyz) [inline object/asset.go]: Determine null outside of it

File: proxy/client/client.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"io"
+	"time"
+
+	"github.com/beego/beego/logs"
```
Comment (hsluoyz) [inline proxy/client/client.go]: We don't use `github.com/beego/beego/logs`, always return error

File: proxy/client/client.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		}
+	}
+}
+
+func (c *Client) sendInitAppMsg(conn *tunnel.Conn) {
+	if c.wantProxyApps == nil {
+		logs.Error("has no app c to proxy")
+	}
+
+	// 通知server开始监听这些app
```
Comment (hsluoyz) [inline proxy/client/client.go]: All English

File: proxy/tunnel/message.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+// limitations under the License.
+
+package tunnel
+
+import "encoding/json"
+
+type MessageType string
+
+const (
+	TypeServerHeartbeat MessageType = "1001"
```
Comment (hsluoyz) [inline proxy/tunnel/message.go]: Don't encode types. Just use descriptive string

File: web/src/AssetEditPage.js [hunk 5]
```diff
@@ -166,11 +168,16 @@ class AssetEditPage extends React.Component {
       assset.services = [];
       assset.enableRemoteApp = false;
     }
+
+    if (!this.state.isIntranet) {
+      assset.hostname = "";
```
Comment (hsluoyz) [inline web/src/AssetEditPage.js]: asset, not assset

File: web/src/AssetEditPage.js [hunk 6]
```diff
@@ -28,6 +28,7 @@ class AssetEditPage extends React.Component {
       assetOwner: props.match.params.organizationName,
       assetName: props.match.params.assetName,
       asset: null,
+      isIntranet: false,
```
Comment (hsluoyz) [inline web/src/AssetEditPage.js]: Remove isIntranet

File: web/src/AssetEditPage.js [hunk 7]
```diff
@@ -262,6 +269,44 @@ class AssetEditPage extends React.Component {
...
+                    this.updateAssetField("hostname", e.target.value);
+                  }} />
+                </Col>
+              </Row>
+              <Row style={{marginTop: "20px"}} >
+                <Col style={{marginTop: "5px"}} span={(Setting.isMobile()) ? 22 : 2}>
+                  {Setting.getLabel(i18next.t("Asset:Remote hostname"), i18next.t("Asset:Remote hostname - Tooltip"))} :
+                </Col>
+                <Col span={22} >
+                  <Input value={asset.remoteHostname || Setting.ServerUrl} onChange={e => {
```
Comment (hsluoyz) [inline web/src/AssetEditPage.js]: Why `|| Setting.ServerUrl`? No need

File: web/src/AssetEditPage.js [hunk 8]
```diff
@@ -262,6 +269,44 @@ class AssetEditPage extends React.Component {
...
+            });
+          }}>{i18next.t("general:Intranet")}
+          </Switch>
+        </Row>
+        {
+          isIntranet && (
+            <React.Fragment>
+              <Row style={{marginTop: "20px"}} >
+                <Col style={{marginTop: "5px"}} span={(Setting.isMobile()) ? 22 : 2}>
+                  {Setting.getLabel(i18next.t("Asset:Hostname"), i18next.t("Asset:Hostname - Tooltip"))} :
```
Comment (hsluoyz) [inline web/src/AssetEditPage.js]: Remove asset.hostname, use asset.name is enough

File: proxy/starter.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/casvisor/casvisor/conf"
+	"github.com/casvisor/casvisor/object"
+	"github.com/casvisor/casvisor/proxy/client"
+	"github.com/casvisor/casvisor/proxy/server"
+	"github.com/casvisor/casvisor/proxy/tunnel"
+	"github.com/casvisor/casvisor/util"
+)
+
+func StartProxyServer() {
+	proxyPort := util.ParseInt(conf.GetConfigString("proxyPort"))
```
Comment (hsluoyz) [inline proxy/starter.go]: GetConfigInt

File: proxy/tunnel/conn.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	return message, nil
+}
+
+// Bind will block until connection close
+func Bind(c1 *Conn, c2 *Conn) {
+	var wait sync.WaitGroup
+	pipe := func(to *Conn, from *Conn) {
+		// 链接断开或发生异常时断开
```
Comment (hsluoyz) [inline proxy/tunnel/conn.go]: All English

File: web/src/AssetEditPage.js (hunk 7 above)
Comment (hsluoyz) [inline web/src/AssetEditPage.js]: If I change the `remoteHostname`'s value now, will it work immediately? Or need to restart Go backend?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: conf/app.conf [hunk 11]
```diff
@@ -17,3 +17,5 @@ clientId = b108dacba027db36ec26
 clientSecret = 124140638b4f9de7e78e79ba22d451c17bfa9688
 casdoorOrganization = "casbin"
 casdoorApplication = "app-casvisor"
+gatewayEndpoint = localhost:18000
+startMode =
```
Comment (hsluoyz) [inline conf/app.conf]: Remove startMode

File: controllers/tunnel.go [hunk 12]
```diff
@@ -275,8 +276,14 @@ func setConfig(propertyMap map[string]string, asset *object.Asset, configuration
 		configuration.Protocol = "vnc"
 	}
 
-	configuration.SetParameter("hostname", asset.Endpoint)
-	configuration.SetParameter("port", strconv.Itoa(asset.Port))
+	if asset.RemotePort != 0 {
+		configuration.SetParameter("hostname", conf.GatewayAddr.IP.String())
```
Comment (hsluoyz) [inline controllers/tunnel.go]: GatewayAddr can be nil

File: proxy/starter.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/casvisor/casvisor/conf"
+	"github.com/casvisor/casvisor/object"
+	"github.com/casvisor/casvisor/proxy/client"
+	"github.com/casvisor/casvisor/proxy/server"
+	"github.com/casvisor/casvisor/proxy/tunnel"
+	"github.com/casvisor/casvisor/util"
+)
+
+func StartProxyServer() {
+	proxyServer, err := server.NewProxyServer("Casvisor Proxy Server", conf.GatewayAddr.Port)
```
Comment (hsluoyz) [inline proxy/starter.go]: GatewayAddr can be nil

File: proxy/starter.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+	if err != nil {
+		panic(fmt.Errorf("failed to get asset by hostname %s", err))
+	}
+	if asset == nil {
+		panic("asset not found")
+	}
+
+	for {
+		client.NewClient(asset.Name,
+			conf.GatewayAddr.IP.String(),
```
Comment (hsluoyz) [inline proxy/starter.go]: GatewayAddr can be nil

File: proxy/starter.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+		panic(fmt.Errorf("failed to create proxy server %s", err))
+		return
+	}
+	proxyServer.Serve()
+}
+
+func StartProxyClient() {
+	asset, err := object.GetAssetByName(util.GetHostname())
+	if err != nil {
+		panic(fmt.Errorf("failed to get asset by hostname %s", err))
```
Comment (hsluoyz) [inline proxy/starter.go]: panic(err)

File: proxy/starter.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+	for {
+		client.NewClient(asset.Name,
+			conf.GatewayAddr.IP.String(),
+			conf.GatewayAddr.Port,
+			tunnel.AssetToAppInfo(asset),
+		).Run()
+	}
+}
+
+func StartMode() string {
```
Comment (hsluoyz) [inline proxy/starter.go]: Remove StartMode

File: web/src/AssetEditPage.js [hunk 17]
```diff
@@ -83,6 +83,7 @@ class AssetEditPage extends React.Component {
 
           this.setState({
             asset: newAsset,
+            isIntranet: newAsset.hostname !== "",
```
Comment (hsluoyz) [inline web/src/AssetEditPage.js]: Remove isIntranet

File: web/src/AssetEditPage.js [hunk 18]
```diff
@@ -153,24 +154,24 @@ class AssetEditPage extends React.Component {
...
+      asset.services = [];
+      asset.enableRemoteApp = false;
     }
-    return assset;
+    return asset;
   }
 
   renderAsset() {
-    const asset = this.state.asset;
+    const {asset, isIntranet} = this.state;
```
Comment (hsluoyz) [inline web/src/AssetEditPage.js]: Remove isIntranet

File: web/src/AssetEditPage.js [hunk 19]
```diff
@@ -262,6 +263,17 @@ class AssetEditPage extends React.Component {
             </Col>
           </Row>
         }
+        <Row style={{marginTop: "20px"}} >
+          <Col style={{marginTop: "5px"}} span={(Setting.isMobile()) ? 22 : 2}>
+            {Setting.getLabel(i18next.t("general:Intranet"), i18next.t("general:Intranet - Tooltip"))} :
+          </Col>
+          <Switch checked={isIntranet} onChange={checked => {
```
Comment (hsluoyz) [inline web/src/AssetEditPage.js]: Remove isIntranet

File: conf/conf.go [hunk 20]
```diff
@@ -15,13 +15,27 @@
...
 
 	"github.com/beego/beego"
+	"github.com/casvisor/casvisor/util"
 )
 
+var GatewayAddr *net.TCPAddr
+
+func init() {
+	var err error
+	GatewayAddr, err = net.ResolveTCPAddr("tcp", GetConfigString("gatewayEndpoint"))
```
Comment (hsluoyz) [inline conf/conf.go]: If gatewayEndpoint is "" (empty string), means disable it

File: conf/conf.go [hunk 21]
```diff
@@ -15,13 +15,27 @@
...
+	if err != nil {
+		fmt.Println("parse gatewayAddr failed")
```
Comment (hsluoyz) [inline conf/conf.go]: If gatewayEndpoint is not empty string, and failed, then panic

File: main.go [hunk 22]
```diff
@@ -20,6 +20,8 @@ import (
 	_ "github.com/beego/beego/session/redis"
 	"github.com/casvisor/casvisor/authz"
 	"github.com/casvisor/casvisor/object"
+	"github.com/casvisor/casvisor/proxy"
+
```
Comment (hsluoyz) [inline main.go]: Remove empty line

File: object/asset.go [hunk 23]
```diff
@@ -284,6 +303,7 @@ func AssetHook(asset *Asset, oldAsset *Asset, action string) error {
 		if err != nil {
 			return err
 		}
+
```
Comment (hsluoyz) [inline object/asset.go]: No new line

File: proxy/client/client.go (hunk 2 above)
Comment (hsluoyz) [reply proxy/client/client.go]: @leo220yuyaodog not fixed

File: object/asset.go [hunk 24]
```diff
@@ -130,6 +131,24 @@ func GetAsset(id string) (*Asset, error) {
 	return getAsset(owner, name)
 }
 
+func GetAssetByName(name string) (*Asset, error) {
```
Comment (hsluoyz) [inline object/asset.go]: Use `GetAsset()`, don't create new API, the owner param is the Casdoor org, you can get it

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by leo220yuyaodog** [comment]
@hsluoyz ready for review

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**